File: src/main/java/space/maxkonkin/nasapicbot/service/NasaService.java
package space.maxkonkin.nasapicbot.service;

import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import space.maxkonkin.nasapicbot.client.NasaApiClient;
import space.maxkonkin.nasapicbot.model.LangCode;
import space.maxkonkin.nasapicbot.model.Nasa;
import space.maxkonkin.nasapicbot.model.User;
import space.maxkonkin.nasapicbot.repository.NasaRepository;
import space.maxkonkin.nasapicbot.to.NasaTo;
import space.maxkonkin.nasapicbot.util.NasaUtil;

import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

@Service
@Slf4j
public class NasaService {
    @Setter
    Boolean withTranslate;

    private final NasaApiClient nasaApiClient;
    private final NasaRepository nasaRepository;
    private final TranslateService translateService;

    public NasaService(NasaApiClient nasaApiClient, NasaRepository nasaRepository, TranslateService translateService) {
        this.nasaApiClient = nasaApiClient;
        this.nasaRepository = nasaRepository;
        this.translateService = translateService;
    }

    public NasaTo getToday(User user) {
        try {
            NasaTo to = nasaApiClient.getNASAObject(nasaApiClient.makeNasaApiRequest(""));
            Optional<Nasa> cached = nasaRepository.getByKeys(user.getTranslateLangCode(), to.date());
            if (cached.isPresent()) {
                log.info("Loaded from cache");
                return NasaUtil.getTo(cached.get());
            } else if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
                NasaTo translated = translateService.translateTitleAndExplanation(to, user.getTranslateLangCode());
                nasaRepository.save(NasaUtil.fromTo(translated, user.getTranslateLangCode()));
                log.info("Translated and saved in cache");
                return translated;
            } else {
                return to;
            }
        } catch (IOException e) {
            log.error(e.getMessage());
            return null;
        }
    }

    public NasaTo getRandom(User user) {
        try {
            NasaTo to = nasaApiClient.getNASAObjects(nasaApiClient.makeNasaApiRequest("?count=1"))[0];
            Optional<Nasa> cached = nasaRepository.getByKeys(user.getTranslateLangCode(), to.date());
            if (cached.isPresent()) {
                log.info("Loaded from cache");
                return NasaUtil.getTo(cached.get());
            } else if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
                NasaTo translated = translateService.translateTitleAndExplanation(to, user.getTranslateLangCode());
                nasaRepository.save(NasaUtil.fromTo(translated, user.getTranslateLangCode()));
                log.info("Translated and saved in cache");
                return translated;
            } else {
                return to;
            }
        } catch (IOException | InterruptedException e) {
            log.error(e.getMessage());
            return null;
        }
    }

    public NasaTo getOnDate(LocalDate date, User user) {
        try {
            NasaTo to = nasaApiClient.getNASAObject(nasaApiClient.makeNasaApiRequest("?date=" +
                    date.format(DateTimeFormatter.ISO_LOCAL_DATE)));
            Optional<Nasa> cached = nasaRepository.getByKeys(user.getTranslateLangCode(), to.date());
            if (cached.isPresent()) {
                log.info("Loaded from cache");
                return NasaUtil.getTo(cached.get());
            } else if (withTranslate && user.getTranslateLangCode() != LangCode.EN) {
                NasaTo translated = translateService.translateTitleAndExplanation(to, user.getTranslateLangCode());
                nasaRepository.save(NasaUtil.fromTo(translated, user.getTranslateLangCode()));
                log.info("Translated and saved in cache");
                return translated;
            } else {
                return to;
            }
        } catch (IOException e) {
            log.error(e.getMessage());
            return null;
        }
    }
}

File: src/main/kotlin/space/maxkonkin/nasapicbot/service/NasaService.kt
package space.maxkonkin.nasapicbot.service

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import space.maxkonkin.nasapicbot.client.NasaApiClient
import space.maxkonkin.nasapicbot.model.LangCode
import space.maxkonkin.nasapicbot.model.User
import space.maxkonkin.nasapicbot.repository.NasaRowTableRepository
import space.maxkonkin.nasapicbot.to.NasaTo
import space.maxkonkin.nasapicbot.util.fromTo
import space.maxkonkin.nasapicbot.util.getTo
import java.io.IOException
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Service
class NasaService(
    private val nasaApiClient: NasaApiClient,
    private val nasaRepository: NasaRowTableRepository,
    private val translateService: TranslateService,
    private val withTranslate: Boolean = false
) {
    fun getToday(user: User): NasaTo? {
        return try {
            val to = nasaApiClient.getNASAObject(nasaApiClient.makeNasaApiRequest(""))
            getNasaTo(user, to)
        } catch (e: IOException) {
            log.error(e.message)
            null
        }
    }

    fun getRandom(user: User): NasaTo? {
        return try {
            val to = nasaApiClient.getNASAObjects(nasaApiClient.makeNasaApiRequest("?count=1")).first()
            getNasaTo(user, to)
        } catch (e: Exception) {
            when (e) {
                is IOException, is InterruptedException -> {
                    log.error(e.message)
                    null
                }

                else -> throw e
            }
        }
    }

    fun getOnDate(date: LocalDate, user: User): NasaTo? {
        return try {
            val to = nasaApiClient.getNASAObject(
                nasaApiClient.makeNasaApiRequest(
                    "?date=" +
                            date.format(DateTimeFormatter.ISO_LOCAL_DATE)
                )
            )
            getNasaTo(user, to)
        } catch (e: IOException) {
            log.error(e.message)
            null
        }
    }

    private fun getNasaTo(user: User, to: NasaTo): NasaTo {
        val cached = nasaRepository.getByDateAndLang(
            LocalDate.parse(
                requireNotNull(to.date),
                DateTimeFormatter.ISO_LOCAL_DATE
            ),
            user.translateLangCode
        )
        return if (cached != null) {
            log.info("Loaded from cache")
            getTo(cached)
        } else if (withTranslate && user.translateLangCode != LangCode.EN) {
            val translated = translateService.translateTitleAndExplanation(to, user.translateLangCode)
            nasaRepository.save(fromTo(translated, user.translateLangCode))
            log.info("Translated and saved in cache")
            translated
        } else {
            to
        }
    }
}

private val log = LoggerFactory.getLogger(NasaService::class.java)